message/validation/consensus_validation.go

```diff
@@ -0,0 +1,428 @@
+		return consensusDescriptor, msgSlot, err
+	}
+
+	state := mv.consensusState(msg.GetID())
+	for _, signer := range signedMsg.Signers {
+		if err := mv.validateSignerBehavior(state, signer, share, msg); err != nil {
+			return consensusDescriptor, msgSlot, fmt.Errorf("bad signer behavior: %w", err)
+		}
+	}
+
+	if mv.checkSignature {
+		if err := ssvtypes.VerifyByOperators(signedMsg.Signature, signedMsg, mv.netCfg.Domain, spectypes.QBFTSignatureType, share.Committee); err != nil {
+			signErr := ErrInvalidSignature
+			signErr.innerErr = err
+			signErr.got = fmt.Sprintf("domain %v from %v", hex.EncodeToString(mv.netCfg.Domain[:]), hex.EncodeToString(share.ValidatorPubKey))
+			return consensusDescriptor, msgSlot, signErr
+		}
+	}
+
+	for _, signer := range signedMsg.Signers {
+		signerState := state.GetSignerState(signer)
+		if signerState == nil {
+			signerState = state.CreateSignerState(signer)
+		}
+		if msgSlot > signerState.Slot {
+			newEpoch := mv.netCfg.Beacon.EstimatedEpochAtSlot(msgSlot) > mv.netCfg.Beacon.EstimatedEpochAtSlot(signerState.Slot)
+			signerState.ResetSlot(msgSlot, msgRound, newEpoch)
+		} else if msgSlot == signerState.Slot && msgRound > signerState.Round {
+			signerState.ResetRound(msgRound)
+		}
+
+		if mv.hasFullData(signedMsg) && signerState.ProposalData == nil {
+			signerState.ProposalData = signedMsg.FullData
+		}
+
+		signerState.MessageCounts.Record(msg)
+	}
+
+	return consensusDescriptor, msgSlot, nil
+}
+
+func (mv *MessageValidator) validateJustifications(
+	share *ssvtypes.SSVShare,
+	signedMsg *specqbft.SignedMessage,
+) error {
+	pj, err := signedMsg.Message.GetPrepareJustifications()
+	if err != nil {
+		e := ErrMalformedPrepareJustifications
+		e.innerErr = err
+		return e
+	}
+
+	if len(pj) != 0 && signedMsg.Message.MsgType != specqbft.ProposalMsgType {
+		e := ErrUnexpectedPrepareJustifications
+		e.got = signedMsg.Message.MsgType
+		return e
+	}
+
+	rcj, err := signedMsg.Message.GetRoundChangeJustifications()
+	if err != nil {
+		e := ErrMalformedRoundChangeJustifications
+		e.innerErr = err
+		return e
+	}
+
+	if len(rcj) != 0 && signedMsg.Message.MsgType != specqbft.ProposalMsgType && signedMsg.Message.MsgType != specqbft.RoundChangeMsgType {
+		e := ErrUnexpectedRoundChangeJustifications
+		e.got = signedMsg.Message.MsgType
+		return e
+	}
+
+	if signedMsg.Message.MsgType == specqbft.ProposalMsgType {
+		if err := instance.IsProposalJustification(share, rcj, pj, signedMsg.Message.Height, signedMsg.Message.Round, signedMsg.FullData); err != nil {
+			e := ErrInvalidJustifications
+			e.innerErr = err
+			return e
+		}
+	}
+
+	return nil
+}
+
+func (mv *MessageValidator) validateSignerBehavior(
+	state *ConsensusState,
+	signer spectypes.OperatorID,
+	share *ssvtypes.SSVShare,
+	msg *queue.DecodedSSVMessage,
+) error {
+	signedMsg, ok := msg.Body.(*specqbft.SignedMessage)
+	if !ok {
+		panic("validateSignerBehavior should be called on signed message")
+	}
```

Comment:
why not requiring this type in the func parameter instead?

Reply:
@moshe-blox Because I need some data from SSVMessage. Although I see that this data can be passed, so I'll try to get rid of it

Reply:
@moshe-blox Fixed in https://github.com/bloxapp/ssv/pull/1066/commits/ee711ef998274a049d711366c9f0cd49fb49ee40